sdks/python/apache_beam/pipeline.py

```diff
@@ -571,6 +573,10 @@ def _infer_result_type(self, transform, inputs, result_pcollection):
       else:
         result_pcollection.element_type = transform.infer_output_type(
             input_element_type)
+    elif isinstance(result_pcollection, pvalue.DoOutputsTuple):
+      # Single-input, multi-output inference.
+      for pcoll in result_pcollection:
+        self._infer_result_type(transform, inputs, pcoll)
```

Comment:
How does this infer a different type to the different collections?

Reply:
I do not believe it does, I missed this while reviewing.

I have a more fundamental question. Is it possible to infer types for multiple outputs without explicit typehints from the user? DoFn will choose their taggedoutput dynamically based on the element. And I am not sure how we can infer different output types only by knowing the input type and all list of possible return types from a function. In other words, is it possible to infer output types (without explicit hints) to be anything other than union of all possible output types from that transform?

Reply:
This indeed does set the same output type for each pcoll. Maybe there's a way to infer more deeply. I can look at this later if this would be of value (I'm thinking that we should come up with a way to specify multi-output type hints).
Currently this is better that leaving pcoll.element_type as None, which fails type checking.

Reply:
Looks like the best bet right is to hard-code multi-outputs to Any.

Reply:
That would fix the bug that they're currently None, but the TODO would still need to be left in place.

Reply:
Done; thanks for the input!